kotlin/src/test/kotlin/io/smallrye/mutiny/coroutines/DeferredAsUniTest.kt

```diff
@@ -78,4 +76,35 @@ class DeferredAsUniTest {
             assertThat(subscriber.awaitItem().item).isNull()
         }
     }
+
+    @Test
+    fun `test uni cancellation before deferred completes`() {
+        runBlocking {
+            // Given
+            val deferredLatch = CountDownLatch(1)
+            val deferred = GlobalScope.async {
+                delay(10_000) // simulate long running job
+                "Hello, World!"
+            }
+            deferred.invokeOnCompletion {
+                deferredLatch.countDown()
+            }
+            val uniLatch = CountDownLatch(1)
+            val uni = deferred.asUni()
+                .onCancellation().invoke { uniLatch.countDown() }
+
+            // When
+            val subscriber = UniAssertSubscriber.create<String>()
+            uni.subscribe().withSubscriber(subscriber)
+            subscriber.cancel()
+
+            // Then
+            assertThat(uniLatch.await(2, TimeUnit.SECONDS)).`as`("Check that uni is cancelled within 2 seconds")
+                .isTrue()
+            assertThat(deferredLatch.await(2, TimeUnit.SECONDS)).`as`("Check that deferred is cancelled within 2 seconds")
+                .isTrue()
+            assertThat(deferred.isCancelled).`as`("Check that Deferred is cancelled")
+                .isTrue()
+        }
+    }
```

Comment:
can you please add a test to ensure https://github.com/smallrye/smallrye-mutiny/issues/752#issuecomment-969481542 in combination with your implementation for the following scenario (pseudo-code):

```
 runBlocking { // parent coroutine
    val child1 = async { // child#1
        doSomeWork // be at this point, e.g. via a lock mechanism at time when child#2.uni is cancelled
        delay() // this call will check coroutine status, and must succeed even after child#2.uni is cancelled
    }
    
    val uni = async { // child#2
    	awesomeWorkToDo // be at this point at time of cancellation
    	delay() // this call will latest detect cancellation and will fail - but must not course trouble as a cancelled Uni should not propagate failure
    }.asUni()
    
    val child3 = async { // child#3
    	// exactly the same here as in child#1
    }
    
    assertThat children 1,2,3 are active
    uni.cancel()
    continue and let complete child#1 and child#3
    asserThat child#1 and child#3 completed successfully and no exception got propagated at all
}
```

Reply:
I've implemented a basic test for this. Do you want an actual call to a suspending function for "doSomeWork" or is having a long delay sufficient?

Reply:
looks good, only the little remark about not only a single suspend call inside the children but many ones.

Thanks for that contribution!